pkg/kubectl/cmd/auth.go

```diff
@@ -0,0 +1,451 @@
+	if err != nil {
+		return err
+	}
+
+	// Get an idToken and refreshToken by having the user authenticate.
+	fmt.Fprintf(out, "Waiting for in-browser authentication. Hit ctrl-c to exit.\n")
+	idToken, refreshToken, err := authClient.Login()
+	if err != nil {
+		return err
+	}
+
+	// Update the Token and RefreshToken for the user.
+	user.Token = idToken
+	if refreshToken != "" {
+		oidcInfo := user.OIDCInfo
+		if oidcInfo == nil {
+			user.OIDCInfo = &clientcmdapi.OIDCInfo{}
+			oidcInfo = user.OIDCInfo
+		}
+		oidcInfo.RefreshToken = refreshToken
+	}
+
+	// Save changes to the user.
+	err = cmdconfig.ModifyConfig(configAccess, *cfg, false)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GetCurrentContext(cfg *clientcmdapi.Config) (*clientcmdapi.Context, error) {
+	// Determine if there is CurrentContext set
+	if cfg.CurrentContext == "" {
+		return nil, fmt.Errorf("current-context must be set in config file")
+	}
+
+	// Get the current context
+	ctx, ok := cfg.Contexts[cfg.CurrentContext]
+	if !ok {
+		return nil, fmt.Errorf("context %q does not exist", cfg.CurrentContext)
+	}
+
+	return ctx, nil
+}
+
+func GetCurrentAuthInfo(cfg *clientcmdapi.Config) (*clientcmdapi.AuthInfo, error) {
+	ctx, err := GetCurrentContext(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure that it has an AuthInfo
+	if ctx.AuthInfo == "" {
+		return nil, fmt.Errorf("the current-context must have a user set")
+	}
+	user, ok := cfg.AuthInfos[ctx.AuthInfo]
+	if !ok {
+		return nil, fmt.Errorf("user %q does not exist in the configuration", ctx.AuthInfo)
+	}
+	return user, nil
+}
+
+func GetCurrentCluster(cfg *clientcmdapi.Config) (*clientcmdapi.Cluster, error) {
+	ctx, err := GetCurrentContext(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if ctx.Cluster == "" {
+		return nil, errors.New("the current-context must have a cluster set")
+	}
+	cluster, ok := cfg.Clusters[ctx.Cluster]
+	if !ok {
+		return nil, fmt.Errorf("cluster %q does not exist in the configuration", ctx.Cluster)
+	}
+	return cluster, nil
+}
+
+// OIDCAuthClient is an object performs operations related to OpenID Connect Authentication against the API Server.
+type OIDCAuthClient struct {
+	server     url.URL
+	httpClient *http.Client
+}
+
+// NewOIDCAuthClient creates a new OIDCAuthClient for the given cluster.
+func NewOIDCAuthClient(cluster *clientcmdapi.Cluster) (*OIDCAuthClient, error) {
+	if cluster.Server == "" {
+		return nil, fmt.Errorf("cluster has no server set")
+	}
+
+	srvURL, err := url.Parse(cluster.Server)
+	if err != nil {
+		return nil, err
+	}
+
+	httpClient, err := HTTPClientForCluster(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	return &OIDCAuthClient{
+		server:     *srvURL,
+		httpClient: httpClient,
+	}, nil
+}
+
+// ExchangeRefreshToken exchanges the given refresh token for an ID token.
+func (o *OIDCAuthClient) ExchangeRefreshToken(refreshToken string) (string, error) {
+	reqURL := o.server
+	reqURL.Path = oidc.PathExchangeRefreshToken
```

Comment:
Are there mechanisms for doing such discovery already in the API Server that I could employ for this?